data/abilities.ts

```diff
@@ -1575,7 +1588,7 @@ export const Abilities: {[abilityid: string]: AbilityData} = {
 				this.add('-activate', pokemon, 'ability: Hadron Engine');
 			}
 		},
-		onAnyTerrainStart(target, source) {
+		onTerrainChange(target, source) {
 			const pokemon = this.effectState.target;
 			if (pokemon === source) return;
```

Comment:
```suggestion
			if (target === source) return;
```

Although this still fails if the terrain was set manually by the ability holder, so IMHO you should check the effect as well.